core/src/main/java/com/yahoo/oak/BasicChunk.java

```diff
@@ -9,55 +9,503 @@
+    protected final AtomicInteger numOfEntries;
+
+    // for writing the keys into the off-heap
+    final OakSerializer<K> keySerializer;
+    final OakSerializer<V> valueSerializer;
+    // to compare serilized and object keys
+    protected final OakComparator<K> comparator;
 
-    /*-------------- Constants --------------*/
+
+    /*-------------- Chunks fields --------------*/
     enum State {
         INFANT,
         NORMAL,
         FROZEN,
         RELEASED
     }
 
-    /*-------------- Members --------------*/
-    // to compare serilized and object keys
-    protected OakComparator<K> comparator;
     // in split/compact process, represents parent of split (can be null!)
     private final AtomicReference<BasicChunk<K, V>> creator;
     // chunk can be in the following states: normal, frozen or infant(has a creator)
     private final AtomicReference<State> state;
     private final AtomicReference<Rebalancer<K, V>> rebalancer;
     private final AtomicInteger pendingOps;
-    private final int maxItems;
+
     protected AtomicInteger externalSize; // for updating oak's size (reference to one global per Oak size)
     protected final Statistics statistics;
 
-    /*-------------- Constructors and creators --------------*/
-    /**
-     * This constructor is only used internally to instantiate a BasicChunk without a creator and a state.
-     * The caller should set the creator and state before returning the BasicChunk to the user.
-     */
-    protected BasicChunk(int maxItems, AtomicInteger externalSize, OakComparator<K> comparator) {
-        this.maxItems = maxItems;
-        this.externalSize = externalSize;
-        this.comparator = comparator;
+    BasicChunk(OakSharedConfig<K, V> config, int additionalFieldCount, int entriesCapacity) {
+        this.config = config;
+        this.externalSize = config.size;
+        this.valuesMemoryManager = config.valuesMemoryManager;
+        this.keysMemoryManager = config.keysMemoryManager;
+        this.keySerializer = config.keySerializer;
+        this.valueSerializer = config.valueSerializer;
+        this.comparator = config.comparator;
+
+        // +2 for key and value references that always exist
+        this.array = new EntryArrayDirect(entriesCapacity, additionalFieldCount + 2);
+
+        this.numOfEntries = new AtomicInteger(0);
+
         this.creator = new AtomicReference<>(null);
         this.state = new AtomicReference<>(State.NORMAL);
         this.pendingOps = new AtomicInteger();
         this.rebalancer = new AtomicReference<>(null); // to be updated on rebalance
         this.statistics = new Statistics();
     }
 
+    // EntryState is the state of the entry after inspection of the states of its key and value
+    // For Example:
+    // Entry in which a key is allocated, but without a valid value, is considered not existing
+    // Entry marked deleted only off-heap is already considered deleted
+    enum EntryState {
```

Comment:
It is really weird to me that EntryState is part of the chunk. If so, I would move it to a separate enum class. And please add EMPTY state, even though for now it is not going to be used.

Reply:
I don't think it is weird. A chunk consists of an entry array and it wants to describe the state of an entry inside the chunk. It is very specific to chunks so it makes sense to be part of this class.

Reply:
I am suggesting to change the comment and to make it clearer...